### ISSUE TITLE
Revapi does not seems to resolve the right dependency version in some conditions

### ISSUE THREAD

**Issue by tmortagne** [revapi/revapi#171]
* revapi-java 0.20.0
* revapi-maven-plugin 0.11.2

The project https://github.com/xwiki/xwiki-platform/tree/master/xwiki-platform-core/xwiki-platform-skin/xwiki-platform-skin-skinx is failing with:

```
[INFO] java.field.removed: field org.apache.struts.action.Action.defaultLocale @ com.xpn.xwiki.web.JsxAction: Field removed from class.

[INFO] java.method.removed: method javax.sql.DataSource org.apache.struts.action.Action::getDataSource(javax.servlet.http.HttpServletRequest) @ com.xpn.xwiki.web.JsxAction: Method was removed.

[INFO] java.method.removed: method javax.sql.DataSource org.apache.struts.action.Action::getDataSource(javax.servlet.http.HttpServletRequest, java.lang.String) @ com.xpn.xwiki.web.JsxAction: Method was removed.

[INFO] java.method.removed: method void org.apache.struts.action.Action::saveErrors(javax.servlet.http.HttpServletRequest, org.apache.struts.action.ActionErrors) @ com.xpn.xwiki.web.JsxAction: Method was removed.

[INFO] java.field.removed: field org.apache.struts.action.Action.defaultLocale @ com.xpn.xwiki.web.SsxAction: Field removed from class.

[INFO] java.method.removed: method javax.sql.DataSource org.apache.struts.action.Action::getDataSource(javax.servlet.http.HttpServletRequest) @ com.xpn.xwiki.web.SsxAction: Method was removed.

[INFO] java.method.removed: method javax.sql.DataSource org.apache.struts.action.Action::getDataSource(javax.servlet.http.HttpServletRequest, java.lang.String) @ com.xpn.xwiki.web.SsxAction: Method was removed.

[INFO] java.method.removed: method void org.apache.struts.action.Action::saveErrors(javax.servlet.http.HttpServletRequest, org.apache.struts.action.ActionErrors) @ com.xpn.xwiki.web.SsxAction: Method was removed.

[INFO] java.field.removed: field org.apache.struts.action.Action.defaultLocale @ com.xpn.xwiki.web.sx.AbstractSxAction: Field removed from class.

[INFO] java.method.removed: method javax.sql.DataSource org.apache.struts.action.Action::getDataSource(javax.servlet.http.HttpServletRequest) @ com.xpn.xwiki.web.sx.AbstractSxAction: Method was removed.

[INFO] java.method.removed: method javax.sql.DataSource org.apache.struts.action.Action::getDataSource(javax.servlet.http.HttpServletRequest, java.lang.String) @ com.xpn.xwiki.web.sx.AbstractSxAction: Method was removed.

[INFO] java.method.removed: method void org.apache.struts.action.Action::saveErrors(javax.servlet.http.HttpServletRequest, org.apache.struts.action.ActionErrors) @ com.xpn.xwiki.web.sx.AbstractSxAction: Method was removed.
```

It's compared to https://maven.xwiki.org/releases/org/xwiki/platform/xwiki-platform-skin-skinx/11.10/.

Those fields/methods indeed don't exist in Struts 1.3.10 but XWiki 11.10 also depends on Struts 1.3.10.

The only difference I can think of which is related to Struts is that this module also have a dependency on velocity tools 2.0 which itself depends on an older version of struts (1.3.8) and it's not the case anymore with velocity tool 3.0 (and it started to fail with the Velocity upgrade). But those fields/methods don't exist in 1.3.8 either so even if revapi was not taking into account the `<dependencyManagement>` it would not be enough to explain the issue.

Unfortunately I cannot reproduce it locally, only on our Jenkins build (for example https://ci.xwiki.org/blue/organizations/jenkins/XWiki%2Fxwiki-platform/detail/master/2633/pipeline/12 ("quality" step)) so not easy to debug...

**Comment by surli** [comment]
So for information, I'm able to reproduce on my side, not sure why @tmortagne cannot reproduce since we're using same version of Maven/Java. We tried to delete both our versions of struts and xwiki skinx from local repo, but we just still have the same behaviour: I reproduce, and Thomas does not.

You can find attached the revapi analysis logs produced when executed using `mvn clean install -Pquality`, where the quality profile is executing our revapi checks. I only kept the revapi check (still 7 mo of logs), full log was 16Mo. I still have it if you need some more info.

For info I searched all matches for struts-core in those logs to check the version and I only found 1.3.10 and 1.3.8 as mentioned above by Thomas.

[maven-revapi.log](https://github.com/revapi/revapi/files/3949447/maven-revapi.log)

**Comment by metlos** [comment]
What does `mvn dependency:tree` tell you about both the old and the current version of `xwiki-platform-skin-skinx`?

**Comment by surli** [comment]
You can find attached both files: the only version of struts-core in both file is 1.3.10. Now you can see that on 11.10 we also had struts-taglib and struts-tiles, but our backward compatibility are related to struts-core so I don't really think it can be related. As a matter of fact, I also manage to replicate the issue by adding manually those dependencies in our pom. 

[deps-tree-11.10.txt](https://github.com/revapi/revapi/files/3955943/deps-tree-11.10.txt)
[deps-tree-master.txt](https://github.com/revapi/revapi/files/3955944/deps-tree-master.txt)

**Comment by metlos** [comment]
So I think I know the reason, but I don't know the solution yet...

There are 2 things contributing to this problem:
1. Revapi assumes a "flat" classpath and therefore multiple jars defining the same classes cause trouble - as they do in java itself.
1. For reasons that I have not found out, `mvn dependency:tree` doesn't mention `struts:struts:1.2.7` as a compile dep of `sslext:sslext:jar:1.2-0` (which is an indirect dep of `org.xwiki.platform:xwiki-platform-velocity-webapp:11.10`), which Revapi finds.

So because Revapi (most probably incorrectly) includes both `struts:struts:1.2.7` and `struts:struts-core:1.3.10` on the classpath, it sometimes happens that the defs from struts 1.2.7 are used and sometimes not.

This should be fixed/improved of course.

**Comment by surli** [comment]
> For reasons that I have not found out, mvn dependency:tree doesn't mention struts:struts:1.2.7 as a compile dep of sslext:sslext:jar:1.2-0 (which is an indirect dep of org.xwiki.platform:xwiki-platform-velocity-webapp:11.10), which Revapi finds.

I guess it's because velocity-tools excludes explicitely struts in its pom.xml, see: https://repo1.maven.org/maven2/org/apache/velocity/velocity-tools/2.0/velocity-tools-2.0.pom

```
<dependency>
	<groupId>sslext</groupId>
	<artifactId>sslext</artifactId>
	<version>1.2-0</version>
	<exclusions>
		<exclusion>
			<groupId>struts</groupId>
			<artifactId>struts</artifactId>
		</exclusion>
	</exclusions>
</dependency>
```

**Comment by metlos** [comment]
Yep, that's gonna be it. The artifact resolver in revapi currently uses "raw" Aether so this thing apparently escaped me :) I'll try to see if `revapi-maven-utils` could be implemented with `maven-resolver`. That should make it finally consistent with Maven resolution because Maven itself uses maven resolver as well.

**Comment by tmortagne** [comment]
Maven Resolver is actually just a rename of Aether (which was used by Maven too) in practice. Maybe you are missing something in the resolve request but not sure which missing information could explain this exclusion to not be taken into account.

**Comment by metlos** [comment]
Well, Maven Resolver has definitely a slightly different top-level API that Aether I think misses, which I'm hoping would make it easier to correctly take into account excludes, dep management and other things that I think Aether might not be taking into account all by itself.

**Comment by metlos** [comment]
So I **think** I've fixed it but writing a test would be quite hard.. :) would you be able to run a snapshot version of Revapi built from this PR against xwiki and see if it solved the issue or, $DEITY forbid, caused any other issues?

**Comment by metlos** [comment]
The tests weren't that hard in the end, so I am fairly confident it is fixed now...

**Comment by metlos** [comment]
Fixed in `revapi-maven-plugin:0.11.3`.
